src/report.rs

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Intel SGX Enclave report structures.
+
+use crate::parameters::{Attributes, Features, MiscSelect, Xfrm};
+
+use core::fmt::{self, Display};
+
+#[derive(Debug, PartialEq)]
+#[non_exhaustive]
+pub enum ReportError {
+    InvalidMiscSelect(u32),
+    InvalidFeatures(u64),
+    InvalidXfrm(u64),
+}
+
+impl Display for ReportError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ReportError::InvalidMiscSelect(value) => {
+                write!(f, "Invalid misc select: {}", value)
+            }
+            ReportError::InvalidFeatures(value) => {
+                write!(f, "Invalid features: {}", value)
+            }
+            ReportError::InvalidXfrm(value) => {
+                write!(f, "Invalid xfrm: {}", value)
+            }
+        }
+    }
+}
+
+/// The enclave report body.
+///
+/// For more information see the following documents:
+///
+/// [Intel® Software Guard Extensions (Intel® SGX) Data Center Attestation Primitives: ECDSA Quote Library API](https://download.01.org/intel-sgx/dcap-1.0/docs/SGX_ECDSA_QuoteGenReference_DCAP_API_Linux_1.0.pdf)
+///
+/// Table 5, A.4. Quote Format
+///
+/// [Intel® 64 and IA-32 Architectures Software Developer's Manual Volume 3 (3A, 3B, 3C & 3D): System Programming Guide](https://www.intel.com/content/www/us/en/architecture-and-technology/64-ia-32-architectures-software-developer-vol-3d-part-4-manual.html)
+///
+/// Table 38-21. Layout of REPORT
+#[derive(Clone, Debug)]
+#[repr(C)]
+pub struct ReportBody {
+    cpusvn: [u8; 16],
+    miscselect: u32,
```

Comment:
Since we're always reading fields like these as little endian, should we store them as bytes to prevent misuse? (i.e. using these values directly)

Reply:
If I change them to integers it may just make it more confusing to read and it might change the alignment. I am pretty much using the same fields as https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/94e327a8235bc7c85353dc5e0a6dc1564e6c96c6/driver/win/PLE/include/asm/sgx_arch.h#L234-L250

Reply:
@definitelynobody In the source code you linked to, the "packed" attribute means that the struct will have an alignment of 1. So we should use byte arrays rather than integers, which will have the same effect while also being safer.

Reply:
ok